### PR TITLE
[API]Payment Method integrity check

### PR DESCRIPTION
**Pull request by arti0090**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/PaymentMethodEligibilityValidator.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        Assert::notNull($value);
+
+        /** @var OrderInterface $order */
+        Assert::notNull($order = $this->orderRepository->findOneBy(['tokenValue' => $value]));
+
+        /** @var PaymentInterface $orderPayment */
+        Assert::notNull($orderPayment = $order->getPayments()[0]);
+
+        if (!$orderPayment->getMethod()->isEnabled()) {
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/PaymentMethodEligibilityValidator.php:44]: I guess you should check all payments

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/PaymentMethodEligibilityValidator.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Core\Repository\OrderRepositoryInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Webmozart\Assert\Assert;
+
+/** @experimental */
+final class PaymentMethodEligibilityValidator extends ConstraintValidator
```
Comment (Tomanhez) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/PaymentMethodEligibilityValidator.php:24]: I think you should add `spec` test

**Review by AdamKasp — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/PaymentMethodEligibilityValidator.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        /** @var PaymentMethodEligibility $constraint */
+        Assert::isInstanceOf($constraint, PaymentMethodEligibility::class);
+
+        /** @var OrderInterface $order */
+        Assert::notNull($order = $this->orderRepository->findOneBy(['tokenValue' => $command->getOrderTokenValue()]));
+
+        /** @var PaymentInterface $orderPayment */
+        Assert::notNull($orderPayments = $order->getPayments());
+
+        foreach($orderPayments as $orderPayment) {
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/PaymentMethodEligibilityValidator.php:48]:
```suggestion
        /** @var PaymentInterface $payment */
        foreach($order->getPayments() as $payment) {
```

File: src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/PaymentMethodEligibilityValidatorSpec.php [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+use Sylius\Component\Core\Repository\OrderRepositoryInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidatorInterface;
+use Symfony\Component\Validator\Context\ExecutionContextInterface;
+
+final class PaymentMethodEligibilityValidatorSpec extends ObjectBehavior
+{
+    function let(
+        OrderRepositoryInterface $orderRepository
+    ): void {
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/PaymentMethodEligibilityValidatorSpec.php:34]:
```suggestion
    function let(OrderRepositoryInterface $orderRepository): void
    {
```

File: src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/PaymentMethodEligibilityValidatorSpec.php [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        $value->setOrderTokenValue('token');
+
+        $orderRepository->findOneBy(['tokenValue' => 'token'])->willReturn($order);
+
+        $order->getPayments()->willReturn(new ArrayCollection([$payment->getWrappedObject()]));
+
+        $payment->getMethod()->willReturn($paymentMethod);
+
+        $paymentMethod->getName()->willReturn('bank transfer');
+
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/PaymentMethodEligibilityValidatorSpec.php:143]:
```suggestion
```

**Review by GSadee — APPROVED** [review]
But the rebase is needed

File: src/Sylius/Bundle/ApiBundle/Resources/config/validation/CompleteOrder.xml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ For the full copyright and license information, please view the LICENSE
+ file that was distributed with this source code.
+
+-->
+
+<constraint-mapping xmlns="http://symfony.com/schema/dic/constraint-mapping"
+                    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                    xsi:schemaLocation="http://symfony.com/schema/dic/constraint-mapping http://symfony.com/schema/dic/services/constraint-mapping-1.0.xsd">
+
+    <class name="Sylius\Bundle\ApiBundle\Command\Checkout\CompleteOrder">
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/validation/CompleteOrder.xml:18]: This configuration is duplicated in `Order.xml`, move all constraints for `CompleteOrder` here

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/PaymentMethodEligibility.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+
+/** @experimental */
+final class PaymentMethodEligibility extends Constraint
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/PaymentMethodEligibility.php:19]:
```suggestion
final class OrderPaymentMethodEligibility extends Constraint
```

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/PaymentMethodEligibilityValidator.php [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+use Sylius\Bundle\ApiBundle\Command\OrderTokenValueAwareInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Core\Repository\OrderRepositoryInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Webmozart\Assert\Assert;
+
+/** @experimental */
+final class PaymentMethodEligibilityValidator extends ConstraintValidator
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/PaymentMethodEligibilityValidator.php:25]:
```suggestion
final class OrderPaymentMethodEligibilityValidator extends ConstraintValidator
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, @arti0090! :tada:

**Pull request merged**